src/Runner.Listener/SelfUpdater.cs

```diff
@@ -256,6 +258,24 @@ private async Task DownloadLatestRunner(CancellationToken token)
                 }
 
                 // If we got this far, we know that we've successfully downloaded the runner package
+                // Validate Hash Matches if it is provided
+                using (FileStream stream = File.OpenRead(archiveFile))
+                {
+                    if (!String.IsNullOrEmpty(_targetPackage.HashValue))
+                    {
+                        using (SHA256 sha256 = SHA256.Create())
+                        {
+                            byte[] srcHashBytes = await sha256.ComputeHashAsync(stream);
+                            var hash = PrimitiveExtensions.ConvertToHexString(srcHashBytes);
+                            if (hash != _targetPackage.HashValue)
+                            {
+                                // Hash did not match, we can't recover from this, just throw
+                                throw new Exception($"Computed runner hash {hash} did not match expected Runner Hash {_targetPackage.HashValue} for {_targetPackage.Filename}");
```

Comment:
Will we start doing this on github.com ?
What will the runner do after this hash match failed? stuck in an upgrade loop?
We better setup canary runs in Ring 0 to verify every runner release has the right checksum.

Reply:
We spoke offline!